File: neetcode150/09_heap_priority_queue/0621_task_scheduler.cpp
/*
 * =====================================================================
 * LeetCode 621 : Task Scheduler                                    (Medium)
 * https://leetcode.com/problems/task-scheduler/
 * Category   : Heap / Priority queue / greedy
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Given tasks labeled 'A'..'Z' and a cooldown n (same task cannot
 *   run twice within n steps), schedule the tasks so the CPU is busy
 *   as soon as possible — return the minimum number of time units.
 *   Idle slots are allowed.
 * ---------------------------------------------------------------------
 * INTUITION
 *   The tasks with the HIGHEST frequency dominate the schedule: the
 *   only reason to add idle time is to satisfy the cooldown of the
 *   most frequent task. Two clean views:
 *   (a) formula — count of max-frequency tasks determines the core;
 *   (b) simulation — a max-heap + cooldown queue mimics the CPU.
 * ---------------------------------------------------------------------
 * APPROACH 1 · Math formula (O(n), best)
 *   Let m = max frequency, c = how many tasks share it.
 *   frames = (m - 1) * (n + 1) + c  ⇒ answer = max(frames, tasks.size()).
 * APPROACH 2 · Heap simulation (O(26 log 26))
 *   Each second pop the most frequent available task, push it to a
 *   cooldown FIFO, re-add after n seconds. Idle when the heap is empty.
 * ---------------------------------------------------------------------
 * DEEP DIVE · Why the formula works
 *   The busiest task T with frequency m needs at least m slots; its
 *   (m-1) gaps between occurrences each demand n idle/other-task slots.
 *   Filling those gaps with other tasks (c tasks tie with T: they can
 *   fill the LAST slot row) is always possible greedily, so the bound
 *   is exact — the answer is just the larger of this schedule skeleton
 *   and the raw task count. This is why O(1) beats any heap version.
 * ---------------------------------------------------------------------
 * TIME COMPLEXITY : O(1) — one pass over 26 counters.
 * MEMORY COMPLEXITY: O(1).
 * =====================================================================
 */
#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include <array>
#include <algorithm>
#include <chrono>
#include <sys/resource.h>

class Solution_1 {
public:
    int leastInterval(std::vector<char>& tasks, int n) {
        std::array<int, 26> cnt{};
        for (char t : tasks) ++cnt[t - 'A'];
        int m = *std::max_element(cnt.begin(), cnt.end());
        int c = static_cast<int>(std::count(cnt.begin(), cnt.end(), m));
        int frames = (m - 1) * (n + 1) + c;          // skeleton w/ idle gaps
        return std::max(frames, static_cast<int>(tasks.size()));
    }
};

class Solution_2 {
public:
    int leastInterval(std::vector<char>& tasks, int n) {
        std::array<int, 26> cnt{};
        for (char t : tasks) ++cnt[t - 'A'];
        std::priority_queue<int> heap;               // max-heap of counts
        for (int x : cnt) if (x) heap.push(x);
        int time = 0;
        std::queue<std::pair<int,int>> cooldown;     // (count, readyAt)
        while (!heap.empty() || !cooldown.empty()) {
            ++time;
            if (!heap.empty()) {
                int c = heap.top() - 1; heap.pop();  // run one instance
                if (c > 0) cooldown.push({c, time + n});
            }
            if (!cooldown.empty() && cooldown.front().second == time) {
                heap.push(cooldown.front().first);   // cooled down → ready
                cooldown.pop();
            }
        }
        return time;
    }
};

static long long nowUs() {
    return std::chrono::duration_cast<std::chrono::microseconds>(
        std::chrono::steady_clock::now().time_since_epoch()).count();
}
static long memKb() { struct rusage r{}; getrusage(RUSAGE_SELF, &r); return r.ru_maxrss; }

int main() {
    struct Case { std::vector<char> tasks; int n; int expect; };
    const Case cases[] = {
        {{'A','A','A','B','B','B'}, 2, 8},
        {{'A','A','A','B','B','B'}, 0, 6},
        {{'A','A','A','A','A','A','B','C','D','E','F','G'}, 2, 16},
        {{'A','B','C','D'}, 3, 4}
    };
    const long mem0 = memKb();
    auto t0 = nowUs();
    bool allOk = true;
    for (const auto& c : cases) {
        int r1 = Solution_1().leastInterval(const_cast<std::vector<char>&>(c.tasks), c.n);
        int r2 = Solution_2().leastInterval(const_cast<std::vector<char>&>(c.tasks), c.n);
        bool ok = r1 == c.expect && r2 == c.expect;
        allOk = allOk && ok;
        std::cout << "n=" << c.n << " → A1=" << r1 << " A2=" << r2
                  << " (want " << c.expect << ") " << (ok ? "PASS" : "FAIL") << "\n";
    }
    auto t1 = nowUs();
    std::cout << (allOk ? "PASS : all cases" : "FAIL") << " | time: " << (t1 - t0)
              << " us | mem: " << (memKb() - mem0) << " KB\n";
    return 0;
}

File: neetcode150/09_heap_priority_queue/0621_task_scheduler.py
"""
=====================================================================
LeetCode 621 : Task Scheduler                                    (Medium)
https://leetcode.com/problems/task-scheduler/
Category   : Heap / greedy
---------------------------------------------------------------------
PROBLEM
    Tasks labelled 'A'..'Z' must run on one CPU; the SAME task cannot
    run twice within n steps. Return the minimum number of time units
    (idle allowed) to finish all tasks.
---------------------------------------------------------------------
INTUITION
    The most frequent task dictates the skeleton of the schedule; idle
    slots appear only to satisfy its cooldown. Two views: a closed
    formula (O(1)) and a heap simulation (fidelity for complex cases).
---------------------------------------------------------------------
APPROACH 1 · Math formula (best)
    m = max frequency, c = count of tasks with frequency m.
    answer = max((m - 1) * (n + 1) + c, len(tasks)).
APPROACH 2 · Heap simulation
    Max-heap of remaining counts + FIFO cooldown queue; every second
    run the most frequent available task, park it for n seconds.
---------------------------------------------------------------------
DEEP DIVE · Why the formula is exact
    The busiest task T appears m times; between consecutive copies we
    need n other/ idle slots — that core costs (m-1)(n+1) units, and
    the LAST row fits all c tasks that tie with T. Filling the core
    gaps with the remaining tasks is always possible greedily, so the
    lower bound is tight. If tasks alone already fill every slot
    (len(tasks) larger), no idle is needed. O(1) beats O(26 log 26).
---------------------------------------------------------------------
TIME COMPLEXITY : O(1)  — one pass over 26 counters.
MEMORY COMPLEXITY: O(1).
=====================================================================
"""
import heapq, collections, time, tracemalloc


class Solution_1:
    """Closed-form formula."""

    def leastInterval(self, tasks, n):
        cnt = collections.Counter(tasks)
        m = max(cnt.values())
        c = sum(1 for v in cnt.values() if v == m)
        return max((m - 1) * (n + 1) + c, len(tasks))


class Solution_2:
    """Heap + cooldown-queue simulation."""

    def leastInterval(self, tasks, n):
        cnt = collections.Counter(tasks)
        heap = [-v for v in cnt.values()]
        heapq.heapify(heap)
        cooldown = collections.deque()     # (remaining, readyAt)
        t = 0
        while heap or cooldown:
            t += 1
            if heap:
                rem = -heapq.heappop(heap) - 1
                if rem > 0:
                    cooldown.append((rem, t + n))
            if cooldown and cooldown[0][1] == t:
                heapq.heappush(heap, -cooldown.popleft()[0])
        return t


if __name__ == "__main__":
    cases = [(["A", "A", "A", "B", "B", "B"], 2, 8),
             (["A", "A", "A", "B", "B", "B"], 0, 6),
             (["A", "A", "A", "A", "A", "A", "B", "C", "D", "E", "F", "G"], 2, 16),
             (["A", "B", "C", "D"], 3, 4)]
    tracemalloc.start()
    t0 = time.perf_counter()
    ok = True
    for tasks, n, want in cases:
        r1 = Solution_1().leastInterval(tasks[:], n)
        r2 = Solution_2().leastInterval(tasks[:], n)
        ok &= r1 == want and r2 == want
        print(f"n={n} -> A1={r1} A2={r2} (want {want}) {'PASS' if r1 == want and r2 == want else 'FAIL'}")
    dt = time.perf_counter() - t0
    mem = tracemalloc.get_traced_memory()[1] / 1024
    tracemalloc.stop()
    print(f"{'PASS : all cases' if ok else 'FAIL'} | time: {dt * 1e6:.0f} us | mem: {mem:.0f} KB")
